### DISCUSSION TITLE
Inconsistent default option between Radio Buttons and Select Dropdown fields

### DISCUSSION THREAD

**Discussion by swierczek** [ExpressionEngine/ExpressionEngine#3214] [Feature Requests]
`Select Dropdown` fields don't use the first value as the Default when adding a new entry. I think this used to work in a previous EE version - maybe EE5?

Regardless, the `Select Dropdown` currently functions differently than `Radio Buttons`, which automatically "checks" the first radio input.

Radio Buttons:
<img width="266" alt="image" src="https://user-images.githubusercontent.com/2423727/160689774-7b0e6b88-21c2-4e1b-acb1-8e69e39e4562.png">

Select Dropdown:
<img width="276" alt="image" src="https://user-images.githubusercontent.com/2423727/160689814-b3567f24-91a1-4dcb-bf9c-b41e69d7b627.png">

I would expect these to function the same because they're both "select a single option from a list" - ideally, both pre-select the first option as the Default. So instead of `Choose wisely`, it would pre-select "No" in my example.

Note: it didn't seem to matter how the list was defined - whether it was a `Value/Label Pair` or `Populate the menu manually`:

<img width="757" alt="image" src="https://user-images.githubusercontent.com/2423727/160691703-de30b8ba-5d72-44e3-b390-852bfec02517.png">

**Comment by PaulBaileyAtChan** [comment]
Selecting the first value in a list as a default isn't appropriate behaviour. There's no reason to assume in the general case that the first item is a suitable default, and an arbitrary default makes it much more likely that an incorrect item will be inadvertently chosen. For consistency, leaving *both* select and radio items unselected by default is the right approach here, until/unless at some point EE allows a default to be actively defined.

**Comment by murtaugh** [comment]
~~I would agree these should behave the same, but would strongly disagree that they need to automatically select the first option. This would prevent you from using these fields as optional, and they would never be empty.~~

~~Edit: @PaulBaileyAtChan said it better, and the ability to define a default would be a welcome addition.~~

My opinion has changed after some research, [see my comment below](https://github.com/ExpressionEngine/ExpressionEngine/issues/1946#issuecomment-1086133370).

**Comment by swierczek** [comment]
My ideal implementation of this would allow us to define a default option for the field. My suggestion to have the first be default was just to align the functionality between these two field types. I would love to use a required `Select Dropdown` field with a defined default to save our content editors from having to set it to the most common option for every new entry. So maybe the functionality could be "use the first as a default if the field is required"? Or maybe adding another toggle to the field types that would allow "Pre-select the first option by default?" would allow this functionality to be more flexible.

I get that always pre-selecting the first option if the field is required may not be ideal, because a user choosing an option is generally a useful action. And if an option is already chosen for them on page load, they may overlook the field altogether. So I think allowing the field to be configured in a way to use a default (either the first option or a pre-defined option) is a good compromise.

I have absolutely defined fields as `Radio Buttons` in the past _specifically because_ the first one was pre-selected in the CMS, and that was a major time saver for me and other content editors.

**Comment by PaulBaileyAtChan** [comment]
Agree that allowing a default to be defined is a good solution here — in which case defining the first option as the default is just one option among others, and doesn't have any special significance. But in the absence of that, and for consistency, selects and radios should default to completely unselected. Using the fact that EE currently selects the first radio option by default is a hack, not a proper solution.

Having a toggle which sets the first option as the default isn't a good solution, because it doesn't address the general case, which is that *any* option might the best default.

**Comment by swierczek** [comment]
https://xkcd.com/1172/ 😛

**Comment by murtaugh** [comment]
I've been doing a little more reading on this, and it appears that the intent of the HTML specifications is that a set of radio buttons _always has a selected option_; a radio button is meant to represent a boolean choice (even if the choice is non-binary) — something is always being selected.

That said, as far as I can tell, *none* of the current browsers actually handle radio buttons this way.

Having read this, I've changed my opinion on what EE's default behavior for radio buttons should be — if no other default is selected, EE should continue to select the first option, and the ability to define a default option should be added to the options for the field type. 

(I don't know if this has a practical benefit, but defining the default during field creation means that the system doesn't have to do the work of figuring out if there's a default while displaying the fields later.)

I changed my mind for two reasons:

1. Browser makers might decide to follow the specifications tighter in the future, at which point any system that depends on the option of an empty value in a set of radio buttons is going to have a bad day.
2. Without this behavior, a radio button set is just a select box with a different UI, which I thought was a good thing, but now I don't. I think more variety of behaviors is more valuable than more variety of UIs.
_______

W3.org: https://www.w3.org/TR/html401/interact/forms.html#radio

**Comment by PaulBaileyAtChan** [comment]
Using the first option as the default still bothers me, whatever W3 says. The thing about the situation a browser is in is that it's right at the end of the process, so if it's committed to always having a value (per W3), and no information is available, it has to guess, and while selecting the first option *might* be right, it's still just a guess, and more likely to result in the wrong option being chosen passively.

EE is in a position where it absolutely doesn't have to guess, because it can prevent an entry from being submitted if nothing is chosen for a field. There's existing machinery for that — the "Require field?" setting. Does that make it just like a select field which requires a value? Maybe — although I do think the marker of a different UI is significant. But I do think that any guesses about field values should be avoided if at all possible.

Adding some machinery to define a default for a radio set seems obviously a good idea. There's a reason why it probably shouldn't be *required*, though: reusing common fields across different channels is generally to be encouraged, but they might need different defaults in different channels. Defining the default in the field itself makes that very tricky — and defining the default somehow in the channel settings might be way too complicated.

But, either way, I'd suggest fixing the "Require entry?" setting as true for radio fields, and removing it as an option. So:

- Adding the ability to define a default for a radio field is v. good, but it shouldn't be required, because the field might be used across different channels, and the same default might not make sense in all of them.
- In general, suggest a simple solution of fixing the "Require entry?" setting to be true for all radio fields, so that it forces a positive choice — esp. in the absence of a default — in order to avoid having to guess.

**Comment by intoeetive** [comment]
@ajohnson just letting you know about this discussion happening here

**Comment by blue-dreamer** [comment]
I've always used the logic of a Radio group where one option is always selected, and a Select where an option is optional. In that respect the way EE works is fine for me.

I do however think that adding the ability to assign a "default" value would be a positive change. For instance where you had a Radio group you could specify that the last item is selected by default if that suited your needs.

**Comment by litzinger** [comment]
Can confirm, this is definitely regression in EE select menu. They all default to "Choose wisely" now, where as they used to default to the first option in the list. We use the select field heavily, and relied on the first option as being the default option without having to make the field required, and without requiring the content editor to make a decision on things that should be... default. To correct this we'll have to make several template updates so the default value is used if there is no value, or make the field required, which is not ideal.

If someone wants to create a Select menu and have a blank option as their default, then they can declare that when creating the select menu, which is what I guarantee everyone here does when they're hand coding a `<select>` menu in html, because if you don't, it defaults to the first option (which is what the Select field used to do).

`<option value="">Default</option>`

So in EE's Select field, if you want a blank default value and want to set the field to required, then you enter no value into the first key/value pair when configuring the field.

Whatever is decided, it's clear this is a regression, and will force people to either change their site when upgrading to whichever version of EE this change occurred in, and for those who miss it, like we did, content editors start reporting "weird behavior" because we relied on the default value without having to make the field required.

**Comment by PaulBaileyAtChan** [comment]
For what it's worth:

The last change to EE's behaviour here was at my suggestion (in #1172), and was implemented in [EE6.1.2](https://docs.expressionengine.com/latest/installation/changelog.html#version-612-release-october-13-2021). That just brought select sub-fields in grid and file grid fields in line with standalone select fields, which had been using the "Choose wisely" meta-option for quite a few years, I think — it's certainly already there in EE5. So it was really just about being consistent. Because sub-fields in grid and file grid fields are visually smaller and easier to miss, it's *especially* important that those require a positive choice by the user, so I definitely support the change.

At least in this case, I'm definitely quibbling with the description of it as a "regression". The fact that it might require a change in site behaviour in some cases doesn't make it regressive. The change was submitted as a bug report, and accepted as such. It's listed in the change log for anyone to view before they update (although it's true that the update notes might make it clearer that it's a fix that will change behaviour).

But I think all of this discussion collapses to the fact that EE really needs the ability to define proper defaults for fields with multiple options. Defaulting to *any* unspecified option, whether the first or not, is a procedural hack. The right solution is that it should be possible to define a default, so the behaviour (for select fields, at least) would be:

- Start with the default option, if defined.
- Otherwise, start with the "Choose wisely" meta-option (aside: it might also be nice to be able to customise this label on the meta-option).
- In either case, if "Require field?" is set, don't allow submission until/unless an option is chosen.

**Comment by litzinger** [comment]
I agree setting a default option somehow is preferred. I still think it's a regression in that a ton of select fields on our site suddenly started behaving differently after an upgrade. Right or wrong, it's different and introduced side effects. The only clear solution for us is to review a ton of templates and add conditionals for default values or make a bunch of select fields required that were not previously required and it was fine if content editors ignored it because the default option was totally valid.

**Comment by ajohnson** [comment]
Just catching up on this conversation...

@litzinger That's interesting that you're experiencing a difference in behavior with the default value of the Select fieldtype in EE6 vs. EE5. I'm looking at a few EE5 sites and the default on my end is always "Choose wisely" the same as it is in EE6. Might be worth digging into why we're experiencing different defaults. Also wondering why more people haven't brought this up, because I agree, that could be a major pain/problem when upgrading from EE5 to EE6.

Overall though, definitely some strong opinions here on how this could/should work. Obviously we'd need to be very careful changing any of the existing default behavior with Select and Radio fieldtypes as those changes could introduce the type of painful scenarios to existing sites like Brian is describing.

But layering on additional functionality is a safer space. So I think it's worth exploring the idea of allowing a default selected value to be specified for both Select and Radio fieldtypes. Doing so wouldn't address every complaint, but it'd be a big step forward.

And @PaulBaileyAtChan, that's a good point regarding needing a field being used in multiple channels to have a different default value selected for each channel. My gut reaction is that assigning a default selected value to a field is enough of a unique attribute to warrant recreating different versions of that fieldtype to work properly in each channel. Or forego selecting a default value altogether if you know ahead of time the field will be used in multiple channels.

I'll brainstorm some ideas for what the UI could look like for indicating which option should be selected by default. Feel free to post your own ideas if you have them, too.

Thanks!
